rust-runtime/aws-smithy-http/src/callback.rs

```diff
@@ -0,0 +1,172 @@
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        // "Use" bytes so that the compiler won't complain.
+        let _ = bytes;
+        Ok(())
+    }
+
+    /// This callback is called once all chunks have been read. If the callback encountered one or more errors
+    /// while running `update`s, this is how those errors are raised. Implementors may return a [`HeaderMap`][HeaderMap]
+    /// that will be appended to the HTTP body as a trailer. This is only useful to do for streaming requests.
+    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        Ok(None)
+    }
+
+    /// Create a new `BodyCallback` from an existing one. This is called when a `BodyCallback` needs to be
+    /// re-initialized with default state. For example: when a request has a body that needs to be
+    /// rebuilt, all callbacks for that body need to be run again but with a fresh internal state.
+    fn make_new(&self) -> Box<dyn BodyCallback>;
+}
+
+impl BodyCallback for Box<dyn BodyCallback> {
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        self.as_mut().update(bytes)
+    }
+    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        self.as_ref().trailers()
+    }
+    fn make_new(&self) -> Box<dyn BodyCallback> {
+        self.as_ref().make_new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{BodyCallback, BoxError};
+    use crate::body::SdkBody;
+    use crate::byte_stream::ByteStream;
+    use std::sync::atomic::{AtomicUsize, Ordering};
+    use std::sync::Arc;
+
+    #[tracing_test::traced_test]
+    #[tokio::test]
+    async fn callbacks_are_called_for_update() {
+        struct CallbackA;
+        struct CallbackB;
+
+        impl BodyCallback for CallbackA {
+            fn update(&mut self, _bytes: &[u8]) -> Result<(), BoxError> {
+                tracing::debug!("callback A was called");
+
+                Ok(())
+            }
+
+            fn make_new(&self) -> Box<dyn BodyCallback> {
+                Box::new(Self)
+            }
+        }
+
+        impl BodyCallback for CallbackB {
+            fn update(&mut self, _bytes: &[u8]) -> Result<(), BoxError> {
+                tracing::debug!("callback B was called");
+
+                Ok(())
+            }
+
+            fn make_new(&self) -> Box<dyn BodyCallback> {
+                Box::new(Self)
+            }
+        }
+
+        let mut body = SdkBody::from("test");
+        body.with_callback(Box::new(CallbackA))
+            .with_callback(Box::new(CallbackB));
+
+        let body = ByteStream::from(body).collect().await.unwrap().into_bytes();
+        let body = std::str::from_utf8(&body).unwrap();
+
+        // Assert that the body that went in is the same as the body coming out.
+        assert_eq!(body, "test");
+
+        assert!(logs_contain("callback A was called"));
+        assert!(logs_contain("callback B was called"));
```

Comment:
Do we care what order these were called in? I don't think the log testing will inform on that.

Reply:
I think no. I'm uncomfortable with users depending on callbacks running in a specific order because it just seems like that would be a source of bugs/confusion. This test exists to show that all callbacks get called when setting more than one.